File: tools/qingming-adapt/stitch-districts.mjs
#!/usr/bin/env node
/**
 * Stitch upstream district panels into contents/qingming-riverside/raw/background.webp
 * Requires: sharp (devDep) OR run the Python one-shot in plan/40.
 * Prefer: node with sharp if available in monorepo.
 */
import { mkdir, access } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const repo = path.resolve(__dirname, "../..");
const assets = path.join(repo, "third_party/qingming-riverside/assets");
const outDir = path.join(repo, "contents/qingming-riverside/raw");

const W = 2172;
const H = 724;

async function main() {
  let sharp;
  try {
    sharp = (await import("sharp")).default;
  } catch {
    console.error("sharp not installed. Use: pnpm add -Dw sharp");
    console.error("Or regenerate via the Python stitch documented in plan/40.");
    process.exit(1);
  }
  for (const f of ["district-west.webp", "street-empty.webp", "district-east.webp"]) {
    await access(path.join(assets, f));
  }
  await mkdir(outDir, { recursive: true });
  const [west, center, east] = await Promise.all(
    ["district-west.webp", "street-empty.webp", "district-east.webp"].map((f) =>
      sharp(path.join(assets, f)).resize(W, H).ensureAlpha().png().toBuffer(),
    ),
  );
  await sharp({
    create: { width: W * 3, height: H, channels: 4, background: { r: 255, g: 255, b: 255, alpha: 1 } },
  })
    .composite([
      { input: west, left: 0, top: 0 },
      { input: center, left: W, top: 0 },
      { input: east, left: W * 2, top: 0 },
    ])
    .webp({ quality: 82 })
    .toFile(path.join(outDir, "background.webp"));
  console.log("wrote", path.join(outDir, "background.webp"), `${W * 3}x${H}`);
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: tools/qingming-adapt/stitch-districts.py
#!/usr/bin/env python3
"""Stitch west|center|east into contents/qingming-riverside/raw/background.webp without white seam bleed."""
from pathlib import Path
import numpy as np
from PIL import Image

REPO = Path(__file__).resolve().parents[2]
ASSETS = REPO / "third_party/qingming-riverside/assets"
OUT = REPO / "contents/qingming-riverside/raw"
W, H, FEATHER = 2172, 724, 48


def fit(im: Image.Image) -> Image.Image:
    im = im.convert("RGB")
    return im if im.size == (W, H) else im.resize((W, H), Image.Resampling.LANCZOS)


def crossfade(arr: np.ndarray, left: Image.Image, right: Image.Image, seam: int, feather: int = FEATHER) -> None:
    half = feather // 2
    la, ra = np.array(left), np.array(right)
    for i in range(feather):
        x = seam - half + i
        if not (0 <= x < arr.shape[1]):
            continue
        t = (i + 0.5) / feather
        li, ri = x - (seam - W), x - seam
        if not (0 <= li < W and 0 <= ri < W):
            continue
        arr[:, x] = (1 - t) * la[:, li] + t * ra[:, ri]


def main() -> None:
    west = fit(Image.open(ASSETS / "district-west.webp"))
    center = fit(Image.open(ASSETS / "street-empty.webp"))
    east = fit(Image.open(ASSETS / "district-east.webp"))
    canvas = Image.new("RGB", (W * 3, H))
    canvas.paste(west, (0, 0))
    canvas.paste(center, (W, 0))
    canvas.paste(east, (W * 2, 0))
    arr = np.array(canvas, dtype=np.float32)
    crossfade(arr, west, center, W)
    crossfade(arr, center, east, W * 2)
    OUT.mkdir(parents=True, exist_ok=True)
    out = Image.fromarray(np.clip(arr, 0, 255).astype(np.uint8))
    path = OUT / "background.webp"
    out.save(path, "WEBP", quality=85, method=6)
    print(f"wrote {path} {out.size}")


if __name__ == "__main__":
    main()
